vochain/account.go

```diff
@@ -74,64 +76,101 @@ func (a *Account) AddDelegate(addr common.Address) error {
 		if bytes.Equal(addr.Bytes(), d) {
 			a.DelegateAddrs[i] = a.DelegateAddrs[len(a.DelegateAddrs)-1]
 			a.DelegateAddrs = a.DelegateAddrs[:len(a.DelegateAddrs)-1]
+			return nil
 		}
 	}
+	return fmt.Errorf("cannot delete delegate, not found")
 }
 
 // TransferBalance transfers balance from origin address to destination address,
 // and updates the state with the new values (including nonce).
 // If origin address acc is not enough, ErrNotEnoughBalance is returned.
 // If provided nonce does not match origin address nonce+1, ErrAccountNonceInvalid is returned.
-// If isQuery is set to true, this method will only check against the current state (no changes will be stored)
-func (v *State) TransferBalance(from, to common.Address, amount uint64, nonce uint32, isQuery bool) error {
-	var accFrom, accTo Account
-	if !isQuery {
-		v.Tx.Lock()
-		defer v.Tx.Unlock()
+func (v *State) TransferBalance(from, to common.Address, amount uint64, nonce uint64) error {
+	var accFrom, accTo *Account
+	var err error
+	if accFrom, err = v.GetAccount(from, false); err != nil {
+		return err
+	}
+	if accFrom == nil {
+		return ErrAccountNotFound
 	}
-	accFromRaw, err := v.Tx.DeepGet(from.Bytes(), AccountsCfg)
+	if accTo, err = v.GetAccount(to, false); err != nil || accTo == nil {
+		return err
+	}
+	if accTo == nil {
+		return ErrAccountNotFound
+	}
+	transferCost, err := v.TxCost(models.TxType_SEND_TOKENS, false)
 	if err != nil {
 		return err
 	}
-	if err := accFrom.Unmarshal(accFromRaw); err != nil {
+	if err := accFrom.Transfer(accTo, amount, transferCost, uint32(nonce)); err != nil {
 		return err
 	}
-	accToRaw, err := v.Tx.DeepGet(to.Bytes(), AccountsCfg)
-	if err != nil && !errors.Is(err, arbo.ErrKeyNotFound) {
+	if err := v.SetAccount(from, accFrom); err != nil {
 		return err
-	} else if err == nil {
-		if err := accTo.Unmarshal(accToRaw); err != nil {
-			return err
-		}
 	}
-	if err := accFrom.Transfer(&accTo, amount, nonce); err != nil {
+	if err := v.SetAccount(to, accTo); err != nil {
 		return err
 	}
-	if !isQuery {
-		af, err := accFrom.Marshal()
-		if err != nil {
-			return err
-		}
-		if err := v.Tx.DeepSet(from.Bytes(), af, AccountsCfg); err != nil {
-			return err
-		}
-		at, err := accTo.Marshal()
-		if err != nil {
-			return err
-		}
-		if err := v.Tx.DeepSet(to.Bytes(), at, AccountsCfg); err != nil {
-			return err
-		}
+	return nil
+}
+
+// CollectFaucet transfers balance from faucet generated package address to collector address,
+// and updates the state with the new values (including nonce).
+func (v *State) CollectFaucet(from, to common.Address, amount uint64, nonce uint64) error {
+	var accFrom, accTo *Account
+	var err error
+	if accFrom, err = v.GetAccount(from, false); err != nil || accFrom == nil {
+		return err
+	}
+	if accFrom == nil {
+		return ErrAccountNotFound
+	}
+	if accTo, err = v.GetAccount(to, false); err != nil {
+		return err
+	}
+	if accTo == nil {
+		return ErrAccountNotFound
+	}
+	if amount == 0 {
```

Comment:
The basic, non expensive checks, should be placed on the begining of the function